docs/specs/3-13-2023-commands-fit-n-finsh.md

```diff
@@ -0,0 +1,71 @@
+
+  Monitor, test, and release your app
+    monitor  	: Monitor a deployed application.
+    pipeline 	: Manage and configure your deployment pipelines.
+
+  Authenticate with Azure
+    login    	: Log in to Azure.
+    logout   	: Log out of Azure.
+
+  Miscellaneous commands
+    config   	: Manage azd configurations (ex: default Azure subscription, location).
+    down     	: Delete Azure resources for an application.
+    env      	: Manage environments.
+    restore  	: Restore application dependencies.
+    template 	: Find and view template details.
+```
+
+The commands are grouped by their scenarios, with most important scenarios at the top. It is not in alphabetical order, rather the order the user is expected to run the commands. If we added `package` for example, would be after `init`.
+
+## Command changes
+
+The following changes are proposed:
+
+- `azd infra create` and `azd infra delete` should be removed, as the functionality already exists in `azd provision` and `azd down`, and users will find the duplication somewhat confusing.
+- The `--output` flag should be moved to `experimental` (hidden and unsupported unless with feature toggle), with the exception of `azd auth token --output json` which must be supported due to credential handlers.
+  - When the feature is hidden, the default output behavior of each command will apply.
+  - This is done as `--output` does not have a documented contract. It prints out `stdout` in JSON, while `stderr` messages also contain JSON-enveloped messages. We might consider splitting this behavior in the future.
+
+Discussion warranted:
+
+- Should we offer `azd restore`? I'm curious if users actually find this useful. We currently use this in our `tasks.json`, which seems useful, but the immediate block directly references `npm`, which defeats the nice abstraction. Excerpt from `tasks.json` below:
+  ```json
+    {
+    "label": "Restore Web",
+    "type": "shell",
+    "command": "azd restore --service web",
+    "presentation": {
+        "reveal": "silent"
+    },
+    "problemMatcher": []
+    },
+    {
+        "label": "Web npm start",
+        "detail": "Helper task--use 'Start Web' task to ensure environment is set up correctly",
+        "type": "shell",
+        "command": "npm run start"
+    }
+  ```
+- Should `azd deploy` deploy only the current service, when the working directory is under a service directory? I personally think this is what users might expect, but I can see an argument against it since it may be implicit behavior.
+- Should `azd deploy --service <service>` be moved to `azd deploy <service>`? I personally feel better about the latter design. We can still allow parsing for `--service`, but remove it in deprecated stage.  If in the future, `azd` supports deploying other things besides services, it could look like: `azd deploy --notAService <notAService name>`.
```

Comment:
We've talked at some point about having infrastructure per service.  In that world, I could have imagined how `azd provision` would learn a `--service` flag that would point at the specific service.

I know you had been thinking about questions like: "should provision be able to take an arbitrary bicep file?" and we have some open issues that ask similar questions.  If we went along these lines, I expect we'd have something like this:

`azd provision` (as you said, this is an alias for `azd provision infra/main.bicep`)
`azd provision <path-to-a-bicep-file>` (provison the bicep file)
`azd provision --service <name>` (provision the infrastructure for the given service?)

or perhaps we don't add `--service` and say "run `azd provision api/infra/main.bicep`

Reply:
From a UX perspective, this is my thinking:
- `azd provision/deploy <target>` - By default, there is a default type that `azd` targets. For `azd deploy`, it is `--service`, and for `azd provision` we get to decide in the future.
- The flag becomes the noun-modifier. If we supported files, `azd provision --file <filepath>`.

I think this is a cleaner UX, as we're making an opinionated default on what the default target is.

Forcing users to type `azd deploy --service <x>` seems rather verbose.

Reply:
`azd deploy` deploying application i.e., all services i..e, api, web
`azd deploy api` deploy api service

Your thought is its already implicit in iterating all services of the application when we do `azd deploy`. We can follow the same patter for `azd deploy <service-name>`?

Was --service added with extensibility in mind for adding more options in future?

Reply:
> azd provision --file filepath

Should we add this to be in parity with azd deploy --service name. Does this give the ability to iteratively call provision and deploy. @ellismg does this help one of the APIM issues which was filed?

Reply:
Thanks for explaining the way you think about this, @weikanglim.  The idea that a command has a natural thing it targets and then using flags to change that target makes sense.  I agree that using an argument instead of a flag here is nicer.

> azd provision --file filepath
>> Should we add this to be in parity with azd deploy --service name. Does this give the ability to iteratively call provision and deploy. @ellismg does this help one of the APIM issues which was filed?
 
We have an issue on this here: #1696.  I think it is a useful primitive to add, but I don't think it's necessary for GA.  I do think that, @weikanglim has shown here with his design that we have a thoughtful way to go add it in the future and a way to think about how we would expose the feature via the command line.

Reply:
I would be fine deprecating the `--service` flag in favor of a parameter where the first parameter is considered to be the service name.

I would not consider any changes here to specify bicep files now as part of the command parameter.  This would cause too much cognitive overhead for novice developers and force them to understand the concept of an infra provider for provisioning.  We should default to be convention based where azd will automatically look for the infra templates in the service.  In the future if we want to introduce a more advanced flag to override this then I could see that happen for our power users.

Reply:
I've captured all the thoughts and suggestions in `#1491` in our private repo. Everyone is tagged in it.